### ISSUE TITLE
Wrong code at -O1/2/3 on x86_64-linux_gnu (recent regression)

### ISSUE THREAD

**Issue by shao-hua-li** [llvm/llvm-project#70509]
clang at -O1/2/3 miscompiled the test.

Bisected to a66051c68a43af39f9fd962f71d58ae0efcf860d, which was committed by @aemerson 

Compiler explorer: https://godbolt.org/z/8sMj8G9z9

```console
% cat a.c
int printf(const char *, ...);
int a, d, e = 35435;
char b;
long c;
unsigned j;
short k, l;
int m(int n) {
  char f[] = {0, 9};
  char *g = f, *i = g;
  long h = n;
  do
    *i++ = h /= 10;
  while (h);
  c = i - g;
  while (g < i)
    b = *g++;
  a = c;
  return c;
}
void o(int n) {
  k = e;
  l = (j = k) > n;
  if (l)
    d = 3;
}
void p() {
  int q = m(17);
  o(q + 65533);
}
int main() {
  p();
  printf("%d\n", d);
}
%
% clang -O0 a.c && ./a.out
3
% clang -O1 a.c && ./a.out
0
%
```

**Comment by aemerson** [comment]
I reverted this in d37b283cdd37 until I have time to investigate.

**Comment by dzaima** [comment]
simplified: https://godbolt.org/z/Gxr3ManrK; as a guess, the `else` branch of the changed `if` expects `!IsAshr` which was no longer true?

**Comment by dtcxzyw** [comment]
Fixed by 2228b35f93256449ca2524ed38dd068b87163efc.
